pkg/prom/instance/instance.go

```diff
@@ -390,6 +393,81 @@ func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer) er
 	return nil
 }
 
+// Update accepts a new Config for the Instance and will dynamically update any
+// running Prometheus components with the new values from Config. Update will
+// return an ErrInvalidUpdate if the Update could not be applied.
+func (i *Instance) Update(c Config) error {
+	// NOTE: you must get a lock on componentMtx first, since initialize will
+	// obtain a lock on cfgMutex and will deadlock if that is already locked
+	// by Update.
+	i.componentMtx.Lock()
+	defer i.componentMtx.Unlock()
+
+	i.cfgMutex.Lock()
+	defer i.cfgMutex.Unlock()
+
+	// It's only (currently) valid to update scrape_configs and remote_write, so
+	// if any other field has changed here, return the error.
+	var err error
+	switch {
+	// This first case should never happen in practice but it's included here for
+	// completions sake.
+	case i.cfg.Name != c.Name:
+		err = fmt.Errorf("name cannot be changed dynamically")
```

Comment:
nit: you could probably re-use the same message.

e.g. `func (fieldName string) error { return fmt.Errorf("%s cannot be changed dynamically", fieldName) }`